From: Takashi Iwai <tiwai@suse.de>
Subject: kABI workaround for hid_debug_list kfifo changes
Patch-mainline: Never, SLE15 only
References: CVE-2019-3819, bsc#1123161

The patch
  patches.drivers/HID-debug-fix-the-ring-buffer-implementation.patch
modifies hid_debug_list struct and leads to the kABI incompatibility.

Work around by the usual trick to move the changed fields to the tail,
and avoid the new inclusion with ifdef.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

---
 drivers/hid/hid-debug.c   |    1 -
 include/linux/hid-debug.h |   10 +++++++++-
 2 files changed, 9 insertions(+), 2 deletions(-)

--- a/include/linux/hid-debug.h
+++ b/include/linux/hid-debug.h
@@ -24,7 +24,9 @@
 
 #ifdef CONFIG_DEBUG_FS
 
+#ifndef __GENKSYMS__
 #include <linux/kfifo.h>
+#endif
 
 #define HID_DEBUG_BUFSIZE 512
 #define HID_DEBUG_FIFOSIZE 512
@@ -41,11 +43,17 @@ void hid_debug_exit(void);
 void hid_debug_event(struct hid_device *, char *);
 
 struct hid_debug_list {
-	DECLARE_KFIFO_PTR(hid_debug_fifo, char);
+	/* the following three fields are merely placeholders for kABI compatibility */
+	char *hid_debug_buf;
+	int head;
+	int tail;
 	struct fasync_struct *fasync;
 	struct hid_device *hdev;
 	struct list_head node;
 	struct mutex read_mutex;
+#ifndef __GENKSYSMS__
+	DECLARE_KFIFO_PTR(hid_debug_fifo, char);
+#endif
 };
 
 #else
--- a/drivers/hid/hid-debug.c
+++ b/drivers/hid/hid-debug.c
@@ -30,7 +30,6 @@
 
 #include <linux/debugfs.h>
 #include <linux/seq_file.h>
-#include <linux/kfifo.h>
 #include <linux/sched/signal.h>
 #include <linux/export.h>
 #include <linux/slab.h>
